File: dashboard/src/lib/auth/parity.test.ts
import { describe, it, expect } from "vitest";
import fs from "fs";
import path from "path";
import { computeWorkspaceRole } from "./compute-workspace-role";
import type {
  RoleBinding,
  DirectGrant,
  AnonymousAccessConfig,
  WorkspaceRole,
} from "@/types/workspace";

/**
 * Shared TS↔Go parity fixture. The SAME JSON file is consumed by the Go
 * pkg/workspaceauth parity_test.go so the two implementations cannot drift.
 */
interface ParityCase {
  name: string;
  roleBindings?: RoleBinding[];
  directGrants?: DirectGrant[];
  anonymousAccess?: AnonymousAccessConfig;
  userGroups: string[];
  userIdentity: string;
  anonymous: boolean;
  expectedRole: WorkspaceRole | "";
}

const fixturePath = path.join(
  __dirname,
  "../../../../pkg/workspaceauth/testdata/parity_cases.json"
);

const cases: ParityCase[] = JSON.parse(fs.readFileSync(fixturePath, "utf-8"));

describe("computeWorkspaceRole parity with Go ComputeRole", () => {
  it("loads the shared fixture", () => {
    expect(cases.length).toBeGreaterThan(0);
  });

  for (const tc of cases) {
    it(tc.name, () => {
      const got = computeWorkspaceRole(
        {
          roleBindings: tc.roleBindings,
          directGrants: tc.directGrants,
          anonymousAccess: tc.anonymousAccess,
          userGroups: tc.userGroups,
          userIdentity: tc.userIdentity,
          isAnonymous: tc.anonymous,
        },
        Date.now()
      );
      const want = tc.expectedRole === "" ? null : tc.expectedRole;
      expect(got).toBe(want);
    });
  }
});

File: pkg/workspaceauth/parity_test.go
package workspaceauth

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"
)

// parityCase mirrors one entry in testdata/parity_cases.json. The SAME file is
// consumed by the dashboard's parity.test.ts so the Go and TS implementations
// can't drift. An empty ExpectedRole means "no access".
type parityCase struct {
	Name            string           `json:"name"`
	RoleBindings    []RoleBinding    `json:"roleBindings"`
	DirectGrants    []DirectGrant    `json:"directGrants"`
	AnonymousAccess *AnonymousAccess `json:"anonymousAccess"`
	UserGroups      []string         `json:"userGroups"`
	UserIdentity    string           `json:"userIdentity"`
	Anonymous       bool             `json:"anonymous"`
	ExpectedRole    Role             `json:"expectedRole"`
}

func TestComputeRoleParity(t *testing.T) {
	data, err := os.ReadFile(filepath.Join("testdata", "parity_cases.json"))
	if err != nil {
		t.Fatalf("read parity fixture: %v", err)
	}

	var cases []parityCase
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatalf("unmarshal parity fixture: %v", err)
	}
	if len(cases) == 0 {
		t.Fatal("parity fixture is empty")
	}

	now := time.Now()
	for _, tc := range cases {
		t.Run(tc.Name, func(t *testing.T) {
			got := ComputeRole(Inputs{
				RoleBindings:    tc.RoleBindings,
				DirectGrants:    tc.DirectGrants,
				AnonymousAccess: tc.AnonymousAccess,
				UserGroups:      tc.UserGroups,
				UserIdentity:    tc.UserIdentity,
				Anonymous:       tc.Anonymous,
			}, now)
			if got != tc.ExpectedRole {
				t.Errorf("ComputeRole = %q, want %q", got, tc.ExpectedRole)
			}
		})
	}
}
